src/app/codegen-data-model/CodegenDataModel_Read.cpp

```diff
@@ -0,0 +1,350 @@
+        return EncodeFromSpan<int8_t>(data, isNullable, encoder);
+    case ZCL_INT16S_ATTRIBUTE_TYPE: // Signed 16-bit integer
+        return EncodeFromSpan<int16_t>(data, isNullable, encoder);
+    case ZCL_INT24S_ATTRIBUTE_TYPE: // Signed 24-bit integer
+        return EncodeFromSpan<OddSizedInteger<3, true>>(data, isNullable, encoder);
+    case ZCL_INT32S_ATTRIBUTE_TYPE: // Signed 32-bit integer
+        return EncodeFromSpan<int32_t>(data, isNullable, encoder);
+    case ZCL_INT40S_ATTRIBUTE_TYPE: // Signed 40-bit integer
+        return EncodeFromSpan<OddSizedInteger<5, true>>(data, isNullable, encoder);
+    case ZCL_INT48S_ATTRIBUTE_TYPE: // Signed 48-bit integer
+        return EncodeFromSpan<OddSizedInteger<6, true>>(data, isNullable, encoder);
+    case ZCL_INT56S_ATTRIBUTE_TYPE: // Signed 56-bit integer
+        return EncodeFromSpan<OddSizedInteger<7, true>>(data, isNullable, encoder);
+    case ZCL_INT64S_ATTRIBUTE_TYPE: // Signed 64-bit integer
+        return EncodeFromSpan<int64_t>(data, isNullable, encoder);
+    case ZCL_SINGLE_ATTRIBUTE_TYPE: // 32-bit float
+        return EncodeFromSpan<float>(data, isNullable, encoder);
+    case ZCL_DOUBLE_ATTRIBUTE_TYPE: // 64-bit float
+        return EncodeFromSpan<double>(data, isNullable, encoder);
+    case ZCL_CHAR_STRING_ATTRIBUTE_TYPE: // Char string
+        return EncodeStringLike<CharSpan, ShortPascalString>(data, isNullable, encoder);
+    case ZCL_LONG_CHAR_STRING_ATTRIBUTE_TYPE:
+        return EncodeStringLike<CharSpan, LongPascalString>(data, isNullable, encoder);
+    case ZCL_OCTET_STRING_ATTRIBUTE_TYPE: // Octet string
+        return EncodeStringLike<ByteSpan, ShortPascalString>(data, isNullable, encoder);
+    case ZCL_LONG_OCTET_STRING_ATTRIBUTE_TYPE:
+        return EncodeStringLike<ByteSpan, LongPascalString>(data, isNullable, encoder);
+    default:
+        ChipLogError(DataManagement, "Attribute type 0x%x not handled", static_cast<int>(metadata->attributeType));
+        return CHIP_IM_GLOBAL_STATUS(UnsupportedRead);
+    }
+}
+
+} // namespace
+
+/// separated-out ReadAttribute implementation (given existing complexity)
+///
+/// Generally will:
+///    - validate ACL (only for non-internal requests)
+///    - Try to read attribute via the AttributeAccessInterface
+///    - Try to read the value from ember RAM storage
+CHIP_ERROR CodegenDataModel::ReadAttribute(const InteractionModel::ReadAttributeRequest & request, AttributeValueEncoder & encoder)
+{
+    ChipLogDetail(DataManagement,
+                  "Reading attribute: Cluster=" ChipLogFormatMEI " Endpoint=%x AttributeId=" ChipLogFormatMEI " (expanded=%d)",
+                  ChipLogValueMEI(request.path.mClusterId), request.path.mEndpointId, ChipLogValueMEI(request.path.mAttributeId),
+                  request.path.mExpanded);
+
+    // ACL check for non-internal requests
+    if (!request.operationFlags.Has(InteractionModel::OperationFlags::kInternal))
```

Comment:
So what is an "internal request" and why would those be going through the TLV interfaces?  Especially because the underlying AttributeAccessInterface logic does not have a concept of "internal request"...

Reply:
The background here is that I am looking to abstract ember away from code. However ember has the concept of internal/external via emberAfReadOrWriteAttribute which is external and emberAfWriteAttribute which is marked as `internal` implicitly from the comments that state:

```
...
* This function will not check to see if the attribute is                                                                                                                                                                                  * writable since the read only / writable characteristic                                                                                                                                                                                  * of an attribute only pertains to external devices writing                                                                                                                                                                               * over the air. Because this function is being called locally 
...
```

I am unsure how to express that better. The way it seems intended (at least in our access cluster functions) is that internal code is allowed to both bypass ACL (timer code for example would not create subject descriptors) and bypass read/write flags (you can write read-only attributes, typically allowing factory data to be read in at boot time, like "write" of a serial number).

Reply:
The right answer for things like that is that clusters should expose APIs to set attributes on them as needed, instead of having a generic API that you are allowed to use to just poke the internals of clusters.

That will also avoid the need to stick the attribute values being written through the pinhole of TLV in the process.

In practice, the _IM_ never calls that API.  It's only the application that does, right?

Reply:
Right, I need some ember compatibility layer that does not depend on ember linkage, so I need to in some way provide a way to say "bypass these rules" because we have code requiring this currently (specifically all Accessors, unsure if others call ember directly, they probably do not because we discourage that).

This is a flag essentially saying "accessor call" (or internal app call). It is also true that IM never calls this, yet I need cluster implementations to be able to call it this way. Either I have a flag (like here) or find some other way, but that was unclear (having a custom subject seemed wrong, especially coupled with a requirement to bypass read/write flags).

This could be done better by not having these functions check ACL or r/w flags at all since metadata provides that, however then I cannot make this a drop-in replacement for ember immediately and I would need to:
  - replace ember-compatibility
  - move ACL and read/write flag check out
  - implement accessors because now it is possible
 
 I can probably still do the above, but the internal flag provides me with a stop-gap where things can be worked on in paralle.

Reply:
OK.  I think we should perhaps document that this is temporary until other things get fixed to be sane?  Again, here should be no direct writing of attributes from random places.  Your setup actually makes that part worse, since right now you can't write things that are handled by AAI, but your API nominally allows that....  So we need something to make sure people don't start abusing it and then we have to keep it forever.

Reply:
Documentation updated and usage of mExpanded vs error codes changed.
It seems only memory usage failures are relevant now (and this is what would prevent us currently from returning just a status code instead of a CHIP_ERROR.